library/Icingadb/Common/Macros.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\Common;
+
+use Icinga\Application\Logger;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+
+trait Macros
+{
+    /**
+     * Caches already processed custom vars query for the given object
+     *
+     * @var array
+     */
+    protected static $loadedCustomVars = [];
+
+    /**
+     * Get the given string with macros being resolved
+     *
+     * @param string       $input  The string in which to look for macros
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function expandMacros(string $input, $object): string
+    {
+        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
+            foreach ($matches[1] as $key => $value) {
+                $newValue = $this->resolveMacro($value, $object);
+                if ($newValue !== $value) {
+                    $input = str_replace($matches[0][$key], $newValue, $input);
+                }
+            }
+        }
+
+        return $input;
+    }
+
+    /**
+     * Resolve a macro based on the given object
+     *
+     * @param string       $macro  The macro to resolve
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function resolveMacro(string $macro, $object): string
+    {
+        if ($object instanceof Host) {
+            $objectType = 'host';
+        } else {
+            $objectType = 'service';
+        }
+
+        $hasVarsPrefix = false;
+        if (preg_match('/^(host|service)[.](vars)[.](.+)/i', $macro, $matches)) {
+            $macroType = $matches[1];
+            $hasVarsPrefix = true;
+        } elseif (preg_match('/^(host|service)[.](.+)/i', $macro, $matches)) {
+            $macroType = $matches[1];
+            $properties = explode('.', $matches[2]);
+        } else {
+            $macroType = $objectType;
+        }
+
+        try {
+            if ($hasVarsPrefix || preg_match('/(^)(vars)[.](.+)/i', $macro, $matches)) {// Custom vars
```

Comment:
The preg_match here should not be required. Adjust your first preg_match so that the object type is optional and detect the macro type then instantly instead of later on again with a complex condition. At least I recognize the third `if` here as that.